### ISSUE TITLE
[Feature Request] New-Object accept types in addition to TypeName

### ISSUE THREAD

**Issue by AikenBM** [PowerShell/PowerShell#7514]
Is there a reason that `New-Object` couldn't accept an actual data type instead of requiring a string representation of the data type?  Currently, when you instance an object with `New-Object`, you need to specify the type of the object with a string representation of the object.

    $Uri = New-Object -TypeName 'System.Uri' -ArgumentList 'http://www.example.com/'

This works perfectly well, but something that I find that I do quite often is use data types directly to take advantage of code completion.  If I want an SqlConnection, for example, I type `[sqlconn`, hit tab for code completion, and PowerShell will automatically give me `[System.Data.SqlClient.SqlConnection]`.  I don't have to look up the namespace, and I can tab through the available types if I need to.

It would be nice if you could just specify that datatype with `New-Object` because it makes it so easy to utilize that code completion, while the `-TypeName` parameter simply requires you to know the whole type or to use code completion with square brackets, and then go back, remove the square brackets, and change it into a string.  More frustrating, you can't easily use code completion at the command line this way.  If you want an SQL connection, you can't type `New-Object -TypeName [SqlConn` and hit tab.  Code completion won't work there because it thinks you're specifying a string parameter.

Imagine if I could just say:

    New-Object -DataType [System.Uri] -ArgumentList 'http://www.example.com/'
    New-Object -DataType [System.Data.SqlClient.SqlConnection] -ArgumentList $ConnectionString
    New-Object -DataType [System.Collections.Generic.HashSet[System.String]]

Where `-DataType` accepts objects objects of `System.RuntimeType`.

Or, if I have an object in a variable and I want a new instance of the same type:

    New-Object -DataType $Uri.GetType() -ArgumentList 'http://www.example.com/'

Granted, you can do things like:

    New-Object -TypeName ([System.Uri]).FullName -ArgumentList 'http://www.example.com/'
    New-Object -TypeName $Uri.GetType().FullName -ArgumentList 'http://www.example.com/'

So perhaps it's not really a worthwhile change, but I have to say that I find it irritating every time I instance an object that it would really be much easier for me to just *specify the data type*, but I've got to supply a string instead.

**Comment by daxian-dbw** [comment]
Why not use the `::new` syntax to create a new object? It's faster than `new-object` and does utilize the type directly. For example:
```
[System.Uri]::new('http://www.example.com/')
[System.Data.SqlClient.SqlConnection]::new($ConnectionString)
[System.Collections.Generic.HashSet[System.String]]::new()
```

**Comment by vexx32** [comment]
Yeah, it's honestly just easier to call the constructor directly. I agree `New-Object` should be able to work like this anyway, but the constructor being both easier and faster make this mostly a non-issue for me.

**Comment by BrucePay** [comment]
@AikenBM  Be aware that you aren't actually passing a type literal when you do  `New-Object [some.type]`, you're passing in the _string_ `"[some-type]"`. This is because arguments are parsed in _argument_ mode where '[' is simply treated as being part of a string. If you wrap the type in parens, forcing _expression_ mode parsing as in
```powershell
New-Object ([System.URI]) -argument http://example.com
```
then you get what you expect. There have been requests in the past to allow the type conversion code for `string` to `Type` to also allow type name strings to include '[',']' so `[type] '[int]'` would work along with with `[type] 'int`' but we felt that it was essentially obfuscating the parsing modes making things overall less clear.

**Comment by thezim** [comment]
I typically do this and get completion at the prompt and in VSCode. ISE if I recall too. Am I missing something here?

```powershell
New-Object System.IO.FileInfo -ArgumentList ~/.bashrc
```

**Comment by vexx32** [comment]
Perhaps it's worth simply modifying `New-Object`'s code specifically to account for possible start and end brackets and act accordingly?

**Comment by AikenBM** [comment]
@daxian-dbw Because [it's apparently community best practice](https://github.com/PoshCode/PowerShellPracticeAndStyle/blob/master/Best-Practices/Building-Reusable-Tools.md#pure-01-use-native-powershell-where-possible).

Also, `::new()` is about as well-documented as `@().ForEach()` and `@().Where()` are.  If we're going to rely on it as the primary method of instancing objects, it ought to be much better documented.  It's only ever used in reference to user defined classes in the documentation.

@BrucePay Ok, so _yes_, there _is_ a reason that New-Object couldn't accept an actual data type instead of requiring a string representation of the data type.  I do see that with this:

```
PS > function Test-DataType ($DataType) { $DataType.NameSpace }
PS > Test-DataType ([int])
System
PS > Test-DataType [int]
PS >
```

@thezim Hm.  That is basically what I'm trying to do and it does not work for me via the command line or with VSCode's PowerShell module (VSCode 1.25.1, PS module 1.8.2).  Maybe something on my system is just not working correctly, however, because it does seem to work for my coworker.  And it does seem to work for me if I RDP to a server and run PowerShell there.  Looks like I will need to reset my environment completely and test it again.

**Comment by AikenBM** [comment]
Alright, so after correcting my system, I do get code completion with `New-Object`.  I am not sure what was not working before, but it does explain a number of other quirks I've noticed for awhile.

I do think the suggestion that @vexx32 put forth makes sense, as it's really not clear why these should work:

    New-Object ([System.URI]) -Argument http://example.com
    New-Object System.URI -Argument http://example.com

And this should not:

    New-Object [System.URI] -Argument http://example.com

Particularly when you get to things like typed collections:

    New-Object System.Collections.Generic.HashSet[String]      ## OK
    New-Object ([System.Collections.Generic.HashSet[String]])  ## OK
    New-Object "System.Collections.Generic.HashSet[String]"    ## OK
    [System.Collections.Generic.HashSet[String]]::new()        ## OK

    New-Object [System.Collections.Generic.HashSet[String]]    ## Error
    New-Object "[System.Collections.Generic.HashSet[String]]"  ## Error

I understand what @BrucePay means about obfuscation of what's going on, but you've _still_ got inconsistent behavior.  I understand this is the compromise you've opted for, but if concern about users not knowing what's going on is the reason it's like it is now, why isn't it in the documentation somewhere?  

It seems particularly odd because the `-as`, `-is`, and `-isNot` operators are completely different:

    "5" -is int     ## Error
    "5" -is [int]   ## OK
    "5" -is ([int]) ## OK
    "5" -is 'int'   ## OK
    "5" -is '[int]' ## Error

It's just very confusing to me from a language semantic perspective.  I don't think either behavior is wrong, and I do understand what's going on.  It's just not at all intuitive that there are two behaviors here.  It's not intuitive what the conventions are and there's no clarification in the documentation why there's a difference in behavior.  Even `about_Type_Operators` doesn't talk about it, which is the closest topic I can find that discusses data types.  I'm actually kind of surprised there isn't an `about_Types` document that explains the semantics and conventions of types.
